qualtran/bloqs/controlled_state_preparation.py

```diff
@@ -0,0 +1,349 @@
+#  Copyright 2023 Google LLC
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      https://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+from typing import Dict, Tuple
+
+import attrs
+import numpy as np
+from numpy.typing import ArrayLike
+
+from qualtran import Bloq, BloqBuilder, SelectionRegister, Signature, SoquetT
+from qualtran.bloqs.arithmetic import Add
+from qualtran.bloqs.basic_gates import OneEffect, OneState, ZeroEffect, ZeroState
+from qualtran.bloqs.basic_gates.rotation import Rx
+from qualtran.bloqs.qrom import QROM
+from qualtran.bloqs.select_and_prepare import PrepareOracle
+
+
+@attrs.frozen
+class ControlledStatePreparationUsingRotations(PrepareOracle):
+    r"""Class that implements controlled state preparation using Ry and Rz rotations from [1]. It
+    does not produce any entangled residual qubits.
+
+    Given a quantum state of which the list of coefficients $c_i$ is known
+    $$
+        |\psi \rangle = \sum_{i=0}^{N-1}c_{i}|i\rangle
+    $$
+    this gate prepares $|\psi\rangle$ from $|0\rangle$ conditioned by a control qubit
+    $$
+        U((|0\rangle + |1\rangle)|0\rangle) = |0\rangle |0\rangle + |1\rangle |\psi\rangle.
+    $$
+
+    Args:
+        n_qubits: number of qubits of the state.
+        rot_reg_size: size of the register that is used to store the rotation angles. Bigger values
+            increase the accuracy of the results.
+        state: tuple of length 2^n_qubits that contains the complex coefficients of the state.
+
+    References:
+        [Trading T-gates for dirty qubits in state preparation and unitary synthesis]
+        (https://arxiv.org/abs/1812.00954).
+            Low, Kliuchnikov, Schaeffer. 2018.
+    """
+
+    n_qubits: int
+    rot_reg_size: int
+    state: Tuple
+    adjoint: bool = False
```

Comment:
I think this has to be called something else so as to not clash with the bloqs adjoint method (see and_bloq which has a special uncompute flag.)

Reply:
I've seen that some classes, such as `ReflectionUsingPrepare`, need the adjoint of a gate. Aren't they using this parameter? How to I change the behavior of the gate then?

Reply:
Also, `PhaseGradientState` has the attribute `adjoint`, in case that is a bug or something.

Reply:
Yeah I think this is something that will go away in the future: see [here for example](https://github.com/quantumlib/Qualtran/pull/583/files#diff-70447a4c3b0e7c3f6d090278b95bddb69e98e009b70a54f62d8732dcc13736f3). If you look at the [And bloq](https://github.com/quantumlib/Qualtran/blob/main/qualtran/bloqs/and_bloq.py#L78) as an example, it defines the flag as uncompute (other places it's been updated to is_adjoint). And then you would have to update the bloq method [adjoint ](https://github.com/quantumlib/Qualtran/blob/main/qualtran/bloqs/and_bloq.py#L89). For the moment because it's not throwing an error you can probably ignore this, and we can clear it up later. Indeed this hasn't been uniformly updated throughout qualtran as you point out.